### ISSUE TITLE
Encoded values as Params?

### ISSUE THREAD

**Issue by 0x777** [nikita-volkov/hasql#48]
``` haskell
list :: Params [(LibPQ.Oid, Maybe ByteString)]
```

It is useful when dealing with dynamically generated queries.

**Comment by nikita-volkov** [comment]
I agree this might be useful to some.

However I'd rather go with the following:

```haskell
raw :: Value (LibPQ.Oid, Maybe ByteString)
```

Then the definition of your `list` function becomes just `contramany raw`.

I won't promise to implement that soon, though. A PR would be welcome.

**Comment by 0x777** [comment]
If we go with `raw`, what does it mean to use functions like [`arrayValue`](http://hackage.haskell.org/package/hasql-0.19.12/docs/Hasql-Encoders.html#v:arrayValue) on `raw`?

**Comment by nikita-volkov** [comment]
Right. Actually it seems like our both variants are impossible. What we can do is something more like this:

```haskell
-- | Given an OID and an according array OID
-- produces an encoder of raw bytes.
raw :: Oid -> Oid -> Value (Maybe ByteString)
```

**Comment by 0x777** [comment]
How about this?
```haskell
rawList :: Params [(LibPQ.Oid, Maybe ByteString)]
rawList = Params (Params.Params $ Op encoder)
  where
    encoder = DList.fromList . map (fmap const)
```

**Comment by 0x777** [comment]
Or rather
```haskell
rawList :: Params [(LibPQ.Oid, Bool -> Maybe ByteString)]
rawList = Params (Params.Params $ Op DList.fromList)
```

**Comment by nikita-volkov** [comment]
Oh, right! This you can do.

Can you describe your use-case though? Maybe we could find something better than exporting such a low-level functionality.

**Comment by 0x777** [comment]
I work on a compiler which converts a JSON based query language to SQL. In a nutshell, 
```haskell
exec :: DatabaseSchema -> JSONQuery -> IO (Either Error Aeson.Value)
```
So, given a query, a Postgres prepared statement is generated, and with the ``DatabaseSchema``, the arguments to this statement are converted to binary format (using encoders from postgresql-binary). This statement is then executed with a thin wrapper written on top of ``postgresql-libpq``.

I would like to replace this wrapper with ``hasql``. The starting point would be to use the ``statement`` function. By the time this function is used, the arguments to the prepared statement are already converted to postgres's binary format. Hence, the requirement for the ``Params [(LibPQ.Oid, Maybe ByteString)]`` encoder.
